pkg/controllers/gitserver/image_test.go

```diff
@@ -39,23 +47,74 @@ func TestReconcileGitServerImage(t *testing.T) {
 		},
 	}
 
-	_, err = r.reconcileGitServerImage(ctx, controllerruntime.Request{}, &resource)
+	dockerClient, err := docker.GetDockerClient()
+	if err != nil {
+		t.Errorf("Getting docker client: %v", err)
+	}
+	defer dockerClient.Close()
+	reader, err := dockerClient.ImagePull(ctx, "docker.io/library/registry:2", types.ImagePullOptions{})
+	defer reader.Close()
+	// blocks until pull is completed
+	io.Copy(os.Stdout, reader)
 	if err != nil {
-		t.Errorf("reconcile error: %v", err)
+		t.Fatalf("failed pulilng registry image: %v", err)
 	}
 
-	if !strings.HasPrefix(resource.Status.ImageID, "sha256") {
-		t.Errorf("Invalid or no Image ID in status: %q", resource.Status.ImageID)
+	waitTimeout := time.Second * 90
+	waitInterval := time.Second * 3
+	// very crude. no guarantee that the port will be available by the time request is sent to docker
+	endTime := time.Now().Add(waitTimeout)
+	for {
+		if time.Now().After(endTime) {
+			t.Fatalf("Timed out waiting for port %d to be available", kind.ExposedRegistryPort)
+		}
+		conn, cErr := net.DialTimeout("tcp", net.JoinHostPort("0.0.0.0", strconv.Itoa(int(kind.ExposedRegistryPort))), time.Second*3)
+		if cErr != nil {
+			break
+		}
+		conn.Close()
+		time.Sleep(waitInterval)
```

Comment:
cosmetics here but pushing the check to a `go func` would be nicer

```suggestion
        ready := make(chan bool)
        endTime := time.After(waitTimeout)
	go func() {
		for {
		   ....
		   if cErr == nil {
		       close(ready)
		   }
		}
	}()

	select {
	case <-timeout:
		err := errors.New("Timeout")
		log.Error(err, "Didn't reconcile Gitea on time.")
		return ctrl.Result{}, err
	case <-ready:
		log.Info("Gitea is ready!")
		resource.Status.GiteaAvailable = true
	}
```

Reply:
I'm using the same format that was used here: https://github.com/cnoe-io/idpbuilder/blob/f6a3d8b5aebc77eae0b72d4ed24343f9174db0a8/pkg/controllers/gitserver/controller_test.go#L120

If we are using timeout, we should use the same logic everywhere or make it reusable. not worth the time effort imo.